### DISCUSSION TITLE
Thoughts on reusing the IT cases.

### DISCUSSION THREAD

**Discussion by irvine0109** [apache/iotdb#3706] [Ideas]
- Background

  Today iotdb has many IT cases, some are in the server module, some are in the cluster module, and some cluster test cases are in the testcontainer module. 

  The test case structure causes developers to sometimes write test cases repeatedly to ensure running correctly of the single-node mode and the cluster mode.

- Suggestion

  - At present, the test cases of single-node mode and cluster mode are separated, but they have intersections, and they can be classified for reuse by using Junit's category annotation.

  - In the testcontainer module, test cases are based on the test environment, but the test environment is relatively fixed, we can use maven to configure multiple environments to test the classified test cases.

Hope to get some reply.

**Comment by jixuan1989** [comment]
> At present, the test cases of single-node mode and cluster mode are separated, but they have intersections, and they can be classified for reuse by using Junit's category annotation.

> In the testcontainer module, test cases are based on the test environment, but the test environment is relatively fixed, we can use maven to configure multiple environments to test the classified test cases.

Can you give some examples?

**Reply by irvine0109** [reply]
OK, I will send an example after reorganization.

**Comment by jixuan1989** [comment]
BTW, as discussed in https://lists.apache.org/thread.html/r435eda40eb6b0c73cafeaae4f6d4f83a4957918c767de439cc1861c5%40%3Cdev.iotdb.apache.org%3E 
Github Discussion is mainly for user-help.

For technical discussion, we'd better to post the discussion on Apache mailing list. 

(AT LEAST, we should write an email to dev@iotdb.apache.org  saying, "hey, I start a discussion about xxxxx on github, the link is http://..... ")

**Reply by HTHou** [reply]
> For technical discussion, we'd better to post the discussion on Apache mailing list.
> 
> (AT LEAST, we should write an email to [dev@iotdb.apache.org](mailto:dev@iotdb.apache.org) saying, "hey, I start a discussion about xxxxx on github, the link is http://..... ")

I suggest we pin this rule at the top of github discussion and post it to mailing list, Slack, confluence, wechat groups, the pined issue #1995 on GitHub and any other community platform.

**Comment by chrisdutz** [comment]
This is exactly the type of discussion that should take place on the mailinglist and not here ... github discussions should only be used for user support.

**Reply by jixuan1989** [reply]
> This is exactly the type of discussion that should take place on the mailinglist and not here ... github discussions should only be used for user support.

Yes.

**Comment by irvine0109** [comment]
I already move this discussion to Jira, the link is [here](https://issues.apache.org/jira/browse/IOTDB-1559#).

Thanks for all relay.
